### PR TITLE
Add ENA taxonomy validation for samples

### PR DESCRIPTION
**Pull request by ke4**
Changes:
- [x] Add tax ID validation
- [x] Add scientific name validation
- [x] Cross check if the given tax id and scientific name are matching

### CODE REVIEW

File: validation/schema.py [hunk 1]
```diff
@@ -5,27 +5,45 @@
...
         return errors
 
+    def taxonomy_validation(self, entity, entity_errors):
+        if 'tax_id' in entity and 'scientific_name' in entity:
+            data_to_validate = {self.TAX_ID_KEY: entity['tax_id'],
+                                self.SCIENTIFIC_NAME_KEY: entity['scientific_name']}
+            taxonomy_validation_result = self.taxonomy_validator.validate_data(data_to_validate)
+            for tax_key, error_message in taxonomy_validation_result.items():
+                if entity_errors:
+                    self.append_value(entity_errors, tax_key, error_message)
```
Comment (MightyAx) [inline validation/schema.py]: can be replaced with:
```
entity_errors.setdefault(tax_key, []).append(error_message)
```

File: validation/schema.py [hunk 2]
```diff
@@ -5,27 +5,45 @@
...
+                else:
+                    entity_errors[tax_key] = error_message
```
Comment (MightyAx) [inline validation/schema.py]: Should this be?
```
entity_errors = {tax_key: [error_message]}
```

File: validation/schema.py [hunk 3]
```diff
@@ -65,3 +83,12 @@ def __translate_to_error(schema_errors: dict) -> dict:
...
         return errors
+
+    @staticmethod
+    def append_value(dict_obj, key, value):
+        if key in dict_obj:
+            if not isinstance(dict_obj[key], list):
+                dict_obj[key] = [dict_obj[key]]
+            dict_obj[key].append(value)
+        else:
+            dict_obj[key] = value
```
Comment (MightyAx) [inline validation/schema.py]: Can be removed by using the setdefault method of the dictionary.

File: validation/taxonomy_validator.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        return \
+            self.__validate(
+                self.ENA_SPECIES_LOOKUP_URL, self.SCIENTIFIC_NAME_KEY, data_to_validate)
+
+    def validate_data(self, data_to_validate: dict):
+        errors = {}
+        if self.TAX_ID_KEY not in data_to_validate:
+            errors[self.TAX_ID_KEY] = "Tax_id field is mandatory."
+        if self.SCIENTIFIC_NAME_KEY not in data_to_validate:
+            errors[self.SCIENTIFIC_NAME_KEY] = "Scientific_name field is mandatory."
```
Comment (MightyAx) [inline validation/taxonomy_validator.py]: To match our current error format this should be:
```
if self.TAX_ID_KEY not in data_to_validate:
    errors[self.TAX_ID_KEY] = ["Tax_id field is mandatory."]
if self.SCIENTIFIC_NAME_KEY not in data_to_validate:
    errors[self.SCIENTIFIC_NAME_KEY] = ["Scientific_name field is mandatory."]
```

File: validation/taxonomy_validator.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        return response.json()
+
+    @staticmethod
+    def append_value(dict_obj, key, value):
+        if key in dict_obj:
+            if not isinstance(dict_obj[key], list):
+                dict_obj[key] = [dict_obj[key]]
+            dict_obj[key].append(value)
+        else:
+            dict_obj[key] = [value]
```
Comment (MightyAx) [inline validation/taxonomy_validator.py]: Use above set default method rather than this.

**Review by MightyAx — CHANGES_REQUESTED** [review]
**I would prefer the Taxonomy Validator be independent of the schema validator.**
so that cli.py CovidExcelUtils class is changed to this:
```
class CovidExcelUtils:
    docker_image = 'dockerhub.ebi.ac.uk/ait/json-schema-validator'
    validation_url = 'http://localhost:3020/validate'
    ena_url = 'https://www.ebi.ac.uk/ena/'
    
    [snip]

    def validate(self):
        try:
            with closing(DockerValidator(self.docker_image, self.validation_url)) as validator:
                self.excel.validate(validator)
        except Exception as error:
            logging.warning(f'Error validating schema, using best guess validation. Detected Error: {error}')
            self.excel.validate(ExcelValidator(self.__file_path))
        self.excel.validate(TaxonomyValidator(ena_url, self.__file_path))
```
**I would also like taxonomy_validator.py to be renamed taxonomy.py since it already lives in the validation module.**

Also, please consider splitting out the the http calls to validate and handling the responses from ena into a new services/ena_taxonomy.py

File: validation/schema.py (hunk 2 above)
Comment (ke4) [reply validation/schema.py]: No, the `error_message` is already a list.

File: validation/schema.py (hunk 1 above)
Comment (ke4) [reply validation/schema.py]: I tried it, but I am afraid it is not doing the same. That's why I used that specific method.

File: validation/schema.py (hunk 3 above)
Comment (ke4) [reply validation/schema.py:94]: noop, I need it

File: validation/schema.py (hunk 1 above)
Comment (MightyAx) [reply validation/schema.py]: Ah, If the error message is already a list you would need **extend**:
```
entity_errors.setdefault(tax_key, []).extend(error_message)
```

File: validation/schema.py (hunk 2 above)
Comment (MightyAx) [reply validation/schema.py]: so then:
`entity_errors.setdefault(tax_key, []).extend(error_message)`

File: validation/schema.py (hunk 1 above)
Comment (ke4) [reply validation/schema.py]: `entity_errors.setdefault(tax_key,` []).extend(error_message)` it works.

File: validation/schema.py (hunk 3 above)
Comment (ke4) [reply validation/schema.py]: Done

File: validation/schema.py (hunk 1 above)
Comment (ke4) [reply validation/schema.py]: Done

File: validation/taxonomy_validator.py (hunk 5 above)
Comment (ke4) [reply validation/taxonomy_validator.py]: Removed

File: validation/taxonomy_validator.py (hunk 4 above)
Comment (ke4) [reply validation/taxonomy_validator.py]: No, that would break the code in `taxonomy_validation` method in `schema.py`

**Review by MightyAx — APPROVED** [review]

**Review by ke4 — COMMENTED** [review]
Looks fine by me

**Pull request merged**